### DISCUSSION TITLE
How do you define generic-like schemas?

### DISCUSSION THREAD

**Discussion by GabenGar** [json-schema-org/community#230] [Q&A] [answered]
Suppose I would like to create an equivalent of this generic type but in schema:
```typescript
interface IAPIResponseSuccess<DataType> {
  ok: true
  data: DataType
}

interface IAPIResponseFailure {
  ok: false
  errors: string[]
}

type IAPIResponse<DataType> = IAPIResponseSuccess<DataType> | IAPIResponseFailure
```
I presume without generics it would look like this:
```json
{
  "$id": "api-response"
  "title": "APIResponse",
  "type": "object",
  "required": ["ok"],
  "additionalProperties": false,
  "oneOf": [
    {
      "properties": {
        "ok": {
           "const": true
         }
      }
    },
    {
      "properties": {
        "ok": {
          "const": false
        },
        "errors": {
          "type": "array",
          "items": {
            "type": "string"
          }
        }
      }
    }
  ]
}
```
And the schema at some endpoint like this (skipping all request/response related properties and only focus on body):
```json
{
  "allOf": [
    {
      "$ref": "api-response"
    },
    "data": {
      "$ref": "some-other-schema"
    }
  ]
}
```
It doesn't exactly match the generic behaviour, but the top level can be processed by a decorator to figure out if it's a failure or a success interface, so not too much boilerplate.
Is this the way to go? Kinda sad this setup doesn't translate back into a generic interface in the code and thus will end up in a million non-generic interface imports.

**Comment by handrews** [comment]
The general schema approach (mix in the specific component of the generic at point-of-use with `allOf` and `$ref`) seems correct.  As far as how it translates into code, you'll need to ask that on the repo of the implementation you're using.

**Reply by GabenGar** [reply]
Well it doesn't depend on implementation, since the way the schema structured is akin to a type union, which is semantically different from a generic type even if their usage might end up being similar.

**Comment by gregsdennis** [comment, marked as answer]
I've actually been considering doing a blog post on this.

This is a perfect application for 2020-12's `$dynamicAnchor` and `$dynamicRef`.

Let's start with the generic definition.

```ts
interface IAPIResponseSuccess<DataType> {
  ok: true
  data: DataType
}
```

For this, we want to create the wrapper without knowing what `DataType` is.

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://my.server/IAPIResponseSuccess",
  "type": "object",
  "$defs": {
    "DataType": {
        "$dynamicAnchor": "DataType",
        "not": true
    }
  },
  "properties": {
    "ok": { "const": true },
    "data": { "$dynamicRef": "#DataType" }
  },
  "required": [ "ok", "data" ]
}
```

So this schema defines an object with an `ok` property that must have a value of `true` and a `data` property that never validates.  That's the `not: true` bit.  We can't just use `false` here because we need the `$dynamicAnchor`.)  Basically this means that you can never have a JSON instance that validates against _only_ the `IAPIResponseSuccess` schema.

Now it gets interesting.  Let's fill in the data.

For simplicity, suppose we want the data to simply be an string: `IAPIResponse<string>`.  For this, we need a new schema that references the generic one above and provides a new `$dynamicAnchor` for `DataType`.

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://my.server/IAPIResponseSuccessOfString",
  "$defs": {
    "DataType": {
        "$dynamicAnchor": "DataType",
        "type": "string"
    }
  },
 "$ref": "https://my.server/IAPIResponseSuccess"
}
```

Then you validate JSON data where you expect `data` to be a string against this schema.

The way this works (and @handrews or @jdesrosiers can probably explain this better) is that your "dynamic scope" starts with the root schema (`IAPIResponseSuccessOfString`).  It defines a `$dynamicAnchor: DataType` and then `$ref`s into the generic definition.  The generic definition also declares `$dynamicAnchor: DataType` but at this  point, we already have that defined, so it's ignored here.  When evaluation hits `$dynamicRef: #DataType`, it uses the first one from the beginning of the dynamic scope, which is the one at the root.

So now this instance will pass:

```json
{
  "ok": true,
  "data": "some string"
}
```

---

To get the union with `IAPIResponseFailure`, use `oneOf` just like you've done.

**Reply by jviotti** [reply]
This is a great explanation of `$dynamicRef` and `$dynamicAnchor`. It should at the very least be added as an example somewhere.

**Reply by handrews** [reply]
Yeah great explanation, @gregsdennis!  If anyone wants a visual of how the dynamic scope works with `$dynamicRef` and `$dynamicAnchor`, you can find one in [this presentation](https://docs.google.com/presentation/d/192uw7T2DftoI4r3Zz1c1acMN2_BwYdZKT3gbuEaG_Yc/edit#slide=id.gfdd23b4381_0_628).

**Reply by GabenGar** [reply]
Is having `$defs` section a requirement for these structures? 
In other words is the first schema below equivalent to the first schema in your post?
```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://my.server/IAPIResponseSuccess",
  "type": "object",
  "properties": {
    "ok": { "const": true },
    "data": { 
      "$dynamicRef": {
        "$dynamicAnchor": "DataType",
        "not": true
      }  
    }
  },
  "required": [ "ok", "data" ]
}
```
I can't wrap my head how the second schema would look inlined so I guess it is a requirement.

**Reply by handrews** [reply]
@GabenGar `$dynamicRef` takes a URI reference, not a schema.  You can't inline the results of `$dynamicRef` as they are not known until runtime.  The step-by-step examples in that slide deck that I linked should make that more clear.  There is also a step-by-step example involving regular `$ref`. `$anchor`. `$id`, and `$defs`, so you can see the different behavior.

In 2020-12, `"$dynamicRef": "#DataType"` requires that there be a `"$dynamicAnchor": "DataType"` in the same [schema resource](https://www.ietf.org/archive/id/draft-bhutton-json-schema-01.html#name-root-schema-and-subschemas-).  This restriction will be removed in the next version of JSON Schema.  So in the next version, you won't need that `$dynamicAnchor` in the first schema, but in this version you do.  As the slide deck illustrates, it's used as an intermediate step in determining the final target of the reference.  It's a bit weird to think about, which is why we're changing it.  This is why you can't inline it.  That `$dynamicAnchor` in the upper schema would only be the reference target if you weren't coming from the other (lower) schema.

In the second (lower) schema, the `$dynamicAnchor` schema under `$defs` is where you end up, because you _started_ evaluation with that schema, went through the `$ref`, and then hit the `$dynamicRef` in the upper schema.

It is generally a good practice to put all reference targets (dynamic or otherwise) in `$defs` if they are not the document root.  It's not required, but it's like putting your re-usable functions in a library.

**Reply by GabenGar** [reply]
@handrews 
I've looked at your slides and didn't understand a thing about `"$dynamicRef"`. This looks like something which requires a pre-existing knowledge/experience to understand. When the dynamic ref is declared as `"data": { "$dynamicRef": "#DataType" }`, is the value of it a shorthand for `#$defs/DataType` or does it refer to the dynamic anchor key `"DataType"`?

**Reply by handrews** [reply]
@GabenGar `$dynamicRef` and `$dynamicAnchor` are like `$ref` and `$anchor`, except that exactly which `$dynamicAnchor` becomes the reference target is calculated at runtime.  Both `$anchor` and `$dynamicAnchor` declare [plain-name fragments](https://www.w3.org/TR/2012/WD-fragid-best-practices-20121025/) (which the JSON Schema spec calls [location-independent identifiers](https://www.ietf.org/archive/id/draft-bhutton-json-schema-01.html#name-defining-location-independe).  So `"$dynamicRef": "#DataType"` points to a `"$dynamicAnchor": "DataType"`  (note that the anchor declaration does not include the `#` - this is analogous to how other formats like HTML declare plain-name fragments).

The example in that presentation shows how the `$dynamicRef` initially resolves to the `$dynamicAnchor` in the same schema resource, but then the dynamic scope is searched and it ultimately resolves to a `$dynamicAnchor` in a different schema resource.

**Reply by jdesrosiers** [reply]
Here are a couple more resources to help you understand dynamic references.

- StackOverflow answer - https://stackoverflow.com/a/69774520/1320693
- Video from ASC 2021 talk - https://youtu.be/GjJpRsVffg0?t=1025

**Reply by GabenGar** [reply]
So do I understand it right that dynamic and static anchors create an implicit namespace on the schema? Is it a requirement for the key of `$dynamicAnchor` declaration within `$defs` to be the same as the value for the anchor? If they are different, does it mean the dynamic anchor in the declaration like this:
```json
"$defs": {
  "key": {
    "$dynamicAnchor": "anchorKey"
  }
}
```
Can be referenced by two different ways i.e. `#$defs/key` and `#anchorKey`?

**Reply by jdesrosiers** [reply]
> So do I understand it right that dynamic and static anchors create an implicit namespace on the schema?

I'm not sure what you mean by that exactly, but that doesn't sound right. Anchors as just a marker labeling a position in the JSON document.

> Is it a requirement for the key of `$dynamicAnchor` declaration within `$defs` to be the same as the value for the anchor?

No, `$defs` has nothing to do with the dynamic anchor declaration. It's just a convenient place to put a schema.

> If they are different, does it mean the dynamic anchor in the declaration like this: [...] Can be referenced by two different ways i.e. `#$defs/key` and `#anchorKey`?

What's technically allowed is a bit complicated, but the only usage that makes sense is for a dynamic reference to reference a dynamic anchor. Other usages end up in the dynamic reference just behaving like a regular reference so you should just use a regular reference instead. In the next release, it's likely that the part that makes the answer to this question complicated will go away and the only thing you will be able to reference from a dynamic reference is a dynamic anchor.

**Reply by GabenGar** [reply]
> I'm not sure what you mean by that exactly, but that doesn't sound right. Anchors as just a marker labeling a position in the JSON document.

Let's say we have schemas `A`, `B` and `C`. `A` and `B` have a dynamic anchor `some-dynamic-ref` declared on them. `C` references both schemas and also declares `some-dynamic-ref`. What is the behaviour of that dynamic ref in `C`?

**Reply by handrews** [reply]
> So do I understand it right that dynamic and static anchors create an implicit namespace on the schema?

Both `$anchor` and `$dynamicAnchor` create URI fragments (like HTML's `id` attribute).  Fragments are relative to a base URI, which is explicitly set by `$id` (and if not set that way, governed by RFC 3986, which usually means the schema's retrieval URI).

The only difference between `$anchor` and `$dynamicAnchor` in draft 2020-12 is that when `$dynamicRef` targets a `$dynamicAnchor`-created fragment, there is special dynamic behavior.  Otherwise, the `$dynamicAnchor` fragment is just a normal URI fragment.

> Let's say we have schemas A, B and C. A and B have a dynamic anchor some-dynamic-ref declared on them. C references both schemas and also declares some-dynamic-ref. What is the behaviour of that dynamic ref in C?

The `$dynamicRef` behavior when targeting `$dynamicAnchor` searches the [dynamic scope](https://www.ietf.org/archive/id/draft-bhutton-json-schema-01.html#name-lexical-scope-and-dynamic-s) and checks for a `$dynamicAnchor`-created URI fragment of the same name as the fragment in the `$dynamicRef` value for each base URI associated with each scope, and resolves to the one associated with the furthest (first-evaluated) active scope.

Therefore it is impossible to answer your question without evaluating an instance with those schemas.  That is the entire point of `$dynamicRef`: to determine the resolution at runtime, unlike normal `$ref` (or HTML's `<a href="...">`, or URI-based links in general).

`$dynamicRef`'s "try this fragment out with multiple base URIs and then choose the farthest one" is not normal URI-referencing behavior, so in the future we might make `$dynamicAnchor` and `$dynamicRef` use something other than URIs so the different behavior is more intuitive.

> Can be referenced by two different ways i.e. `#$defs/key` and `#anchorKey`?

Those are just two different fragment syntaxes.  You can reference them normally.  The only special case is that if `$dynamicRef` references a fragment created by `$dynamicAnchor`, then the additional behavior described a few paragraphs above happens.  If anything else (e.g. `$ref`) references fragments, they just behave like normal fragments no matter how they were created (`$anchor`, `$dynamicAnchor`, or the structure of the JSON document as expressed in a JSON Pointer fragment).

**Reply by GabenGar** [reply]
> Therefore it is impossible to answer your question without evaluating an instance with those schemas. That is the entire point of `$dynamicRef`: to determine the resolution at runtime, unlike normal `$ref` (or HTML's `<a href="...">`, or URI-based links in general).

The declaration for `C` is along the lines of:
```json
{
  "properties": {
    "a": {
      "$ref": "A"
    },
    "b": {
      "$ref": "B"
    }
  }
}
```
So does declaring `some-dynamic-ref` change the value in both keys?

**Reply by handrews** [reply]
I don't even know what you're asking at this point.  If you want to know how something evaluates, and how that changes between using `$dynamic*` vs regular `$ref`/`$anchor`, you need to show:

* the complete schema using `$ref`/`$anchor`
* the complete schema using `$dynamicRef`/`$dynamicAnchor`
* the complete instance or instances being evaluated
